keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,479 @@
+However, it is also possible that there could be issues with Kubernetes components or a misconfiguration. 
+If the eviction of the target pod fails, users should first check if the eviction violates the pod's PDB configuration. 
+If it does violate the PDB, users should manually adjust the replica setting of the owner of the pod to ensure that the eviction no longer violates the PDB configuration. Alternatively, users can manually evict the pod.
+Another thing users can check is whether there is incorrect configuration, such as multiple PodDisruptionBudgets referencing the same Pod.
+Abnormal states of kube-apiserver and controller-manager can also cause eviction failures, so they should also be among the objects that users should check.
+
+
+Another concern may be the overhead of the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+However, the impact won't be significant due to the limitation on the number of pods per node.
+The evictions caused by this may be expensive depending on the number of pods and the requests conflicts and blocking PDBs.
+
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeSelector struct {
+	NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeAffinity struct {
+	// If the affinity requirements specified by this field are not met at
+	// scheduling time, the pod will not be scheduled onto the node.
+	// If the affinity requirements specified by this field cease to be met
+	// at some point during pod execution (e.g. due to an update), the system
+	// will try to eventually evict the pod from its node.
+  // For v1alpha1, evictions are performed by calling the Eviction API.
+  // This may change in later versions.
+	// +optional
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector
+}
+```
+
+Add a controller called `node-affinity-eviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node(excluding mirror pods), checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if `NodeSelector` still match the new node. 
+- If `RequiredDuringSchedulingRequiredDuringExecution` is no loger met, trying to evict the pod.
+
+- For alpha stage:
+  - With each occurrence of a node label update event, a new eviction round is triggered. In each round of eviction, the controller will iterate through each pod on that node to determine whether the pod should be evicted.
+  - It will use Eviction API to evict the pod(The API takes the PDBs into consideration, so we do not have to use the PDBs explicitly here). 
+  - If eviction for one pod fails, We will skip the eviction of the target pod in the current round. In the next eviction round, we will recheck the pod and may trigger eviction based on the results of the check.
```

Comment:
This could cause the failed evictions' pods to become stale for a long time. Is there a resync for the eviction rounds?

Reply:
I will deffer this to sig-scheduling folks. I think not having a resync/backoff could be quite annoying in some cases.

The Eviction API would solve this problems, but we are still discussing it (earliest possible is 1.32).

Reply:
Not retrying would cause confusion and issues to be opened.

Not sure what the taint eviction controller does when there is a failure. Does the Eviction API give any indication of when you should retry? If not, maybe a retry with exponential backoff is enough?

Reply:
Currently, many of the open comments here are about retries, so I suggest we focus our discussion here.

> why would the eviction fail? Because of PDB?

1. Because evictions will violate PDB or there is a misconfiguration(like if multiple PodDisruptionBudgets reference the same Pod).
2.  Because API rate limiting or unhealthy component states.

In each case, I don't think immediate retry would help.

Currently, I'm considering the following approaches to address the issue of pods potentially not being retried for a long time:

1. Create an infinite loop for failed evictions' pods.
I think this would increase the burden on the controller, and I am not inclined towards this approach.

2. Retry with exponential backoff for failed evictions' pods.
The concern with this approach is that there is still a significant probability of eviction failures in recent retries if the PDB is not updated in a timely manner.

3. Make controller listening to the pdb updates/deletes and retrigger a new round of eviction.
This can address the first scenario of eviction failures, which accounts for the majority of the failures.
However, triggering evictions due to PDB updates would require traversing all pods, which could indeed impose a significant burden.
But the cost should be manageable considering that this is the current implementation approach of the descheduler, which involves periodically traversing each pod on every node?

Reply:
> Not sure what the taint eviction controller does when there is a failure. Does the Eviction API give any indication of when you should retry? If not, maybe a retry with exponential backoff is enough?

Not presently, but we're discussing the ability to expand it with such functionality. For now don't rely on anything like that.

Reply:
So is the suggestion not to do retries?

Reply:
Retires will definitely not hurt, even today. In the long run, they hopefully should help. But for now I'd advice some kind of backoff for those retries. Maybe in the evolution we'll actually provide some hints when to retry. That's TBD :sweat_smile:

Reply:
> Can we hold another FIFO queue with a long retry duration, not blocking the eviction API.

We still need to traverse this queue right?  Because the head of the queue may never be evicted (if the PDB is not updated continuously)?
If yes, this seems like path 1 in https://github.com/kubernetes/enhancements/pull/4329#discussion_r1636453639, and the concern is the large memory consumption.

Reply:
I have two designs in my mind:
- the first one is quite simple, we'll have a linked list to store the failed-to-evict Pods, it's a FIFO queue, and we'll pop out Pods from the head of the list every 2s, if the Pod stored in the list for more than 2s, let's retry, or wait for next iteration. Each iteration we'll pop out all  Pods that waiting more than 2s.
- the second one is we'll have a map to store the failed-to-evict Pods, each Pod will be wrapped with a timer, once countdown ends, the Pod will retry, the time can be exponential backoff, 1s, 2s, 4s ...

Considering Pod will never be retried successfully, we may have a limited retry times to avoid  the list/map is too large.

Reply:
- I think we could just use exponential RateLimitingQueue with a big maximum delay
- Have we considered using the descheduler logic + descheduler framework? It is pretty minimal and we could have just one plugin for the node affinity eviction. And have an option to add different mechanisms in the future. https://github.com/kubernetes-sigs/descheduler/blob/1c8ae64726f2de1da8240e297af156ea071a3de3/pkg/descheduler/descheduler.go#L385

Reply:
> Have we considered using the descheduler logic + descheduler framework?

I think ultimately we want current framework with evacuation api, since evacuation api can handle the retry logic.
This should have a better performance than descheduler logic.

Since this is not an urgent KEP, waiting for the evacuation API might be a better choice compared to directly incorporating the descheduler framework into the core Kubernetes and subsequently refactoring it. If we cannot accept the retry logic outlined at present.

Reply:
Eesh. The term “evacuation” is being used in a sense quite different from its common technical meaning (similar to “drain”).